File: 2015/day01/C/day01.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *import_input(char *filename) {
  // These are needed for getlines, as it dynamically resizes the pointer
  char *line = NULL;
  size_t len = 0;
  ssize_t read;

  FILE *fptr = fopen(filename, "r");

  if (fptr != NULL) {
    // getlines only works on POSIXs systems
    read = getline(&line, &len, fptr);
  }

  free(line);
  fclose(fptr);
  return line;
}

int count_floor(char *input) {
  int floor = 0;
  int i;

  for (i = 0; i < strlen(input); i++) {
    if (input[i] == '(') {
      floor++;
    } else if (input[i] == ')') {
      floor--;
    }
  }

  return (floor);
}

int first_basement(char *input) {
  int floor = 0;
  int i;

  for (i = 0; i < strlen(input); i++) {
    if (input[i] == '(') {
      floor++;
    } else if (input[i] == ')') {
      floor--;
    }

    if (floor == -1) {
      return i + 1;
    }
  }
  return -1;
}

int main() {
  printf("Hello world!\n");
  char *input;
  int final_floor;
  int basement_index;

  input = import_input("../input.txt");

  final_floor = count_floor(input);
  basement_index = first_basement(input);

  printf("The Final Floor Santa Lands on is: %i\n", final_floor);
  printf("The index of the basement is: %i", basement_index);

  return 0;
}

File: 2015/day01/python/day01.py
def parse_file(file_name: str) -> str:
    with open(file_name) as file:
        # The input is the first line of the file
        input = file.readline()
    return input


def main():
    input = parse_file("../input.txt")

    floor = 0
    for index, char in enumerate(input):
        if char == "(":
            floor += 1
        elif char == ")":
            floor -= 1

        if floor == -1:
            print(f"The First basement entry is on floor %i", index + 1)

    print(f"The final floor is %i:", floor)


if __name__ == "__main__":
    main()
